### PR TITLE
Shuffle persist event data store functions.

### PR DESCRIPTION
**Pull request by erikjohnston**
The aim here is to get to a stage where we have a `PersistEventStore` that holds all the write methods used during event persistence, so that we can take that class out of the `DataStore` mixin and instansiate it separately. This will allow us to instansiate it on processes other than master, while also ensuring it is only available on processes that are configured to write to events stream.

This is a bit of an architectural change, where we end up with multiple classes per data store (rather than one per data store we have now). We end up having:

1. Storage classes that provide high level APIs that can talk to multiple data stores.
2. Data store modules that consist of classes that must point at the same database instance.
3. Classes in a data store that can be instantiated on processes depending on config.

Thoughts welcome if this is the right approach.

### CODE REVIEW

File: synapse/storage/data_stores/main/events.py [hunk 1]
```diff
@@ -30,21 +31,35 @@
...
 from synapse.types import StateMap, get_domain_from_id
 from synapse.util.frozenutils import frozendict_json_encoder
 from synapse.util.iterutils import batch_iter
 
+# py2 sqlite has buffer hardcoded as only binary type, so we must use it,
+# despite being deprecated and removed in favor of memoryview
+if six.PY2:
+    db_binary_type = six.moves.builtins.buffer
+else:
+    db_binary_type = memoryview
```
Comment (clokep) [inline synapse/storage/data_stores/main/events.py:63]: Should we clean this up while we're here?

**Comment by clokep** [comment]
@erikjohnston and I had a long conversation about this, this PR is really made up of a few steps (I'm paraphrasing / restating Erik for my own understanding):
1. Split off some methods to separate classes (essentially code clean-up).
2. Move methods to the same class as their caller (instead of depending on being combined together via a mixin).
3. Move the `persist_events()` code out of the large mixin and ensure it can only be called via the storage layer (e.g. `hs.get_storage().persistence.persist_events(...)` instead of `hs.get_store().persist_events(...)`).

What it comes down to is that the changes to `synapse/storage/data_stores/__init__.py` and ` synapse/storage/persist_events.py` are probably the most interesting (in terms of the changing architecture). The rest is just code shuffling.

File: synapse/storage/data_stores/main/events.py [hunk 2]
```diff
@@ -94,58 +106,45 @@ def f(self, *args, **kwargs):
...
 
-        hs.get_clock().looping_call(read_forward_extremities, 60 * 60 * 1000)
 
-        def _censor_redactions():
-            return run_as_background_process(
-                "_censor_redactions", self._censor_redactions
-            )
+class PersistEventsStore:
+    """T
+    """
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/events.py:127]: I feel like I also need to finish this docstring....

**Comment by richvdh** [comment]
> We end up having:
> 
>     1. Storage classes that provide high level APIs that can talk to multiple data stores.
> 
>     2. Data store modules that consist of classes that must point at the same database instance.
> 
>     3. Classes in a data store that can be instantiated on processes depending on config.

this sounds sensible to me, though the naming is confusing. In particular `Store` and `Storage` sound like basically the same thing, and we have `DataStore` and `StateGroupDataStore` classes currently, which  ais confused by the existence of the current `DataStore` class, which I *think* falls somewhere between categories 1 and 3 currently.

**Comment by clokep** [comment]
👍 on it being confusing that "store" and "storage" mean different things...

**Comment by erikjohnston** [comment]
> > We end up having:
> > ```
> > 1. Storage classes that provide high level APIs that can talk to multiple data stores.
> > 
> > 2. Data store modules that consist of classes that must point at the same database instance.
> > 
> > 3. Classes in a data store that can be instantiated on processes depending on config.
> > ```
> 
> this sounds sensible to me, though the naming is confusing. In particular `Store` and `Storage` sound like basically the same thing, and we have `DataStore` and `StateGroupDataStore` classes currently, which ais confused by the existence of the current `DataStore` class, which I _think_ falls somewhere between categories 1 and 3 currently.

Yup. Ideally in future we can start splitting the `DataStore` up some more, though that is a bit complicated due to the fact that half the code base uses it directly via `get_datastore()`.

We should probably rename `DataStore` (and other functions), I hadn't quite twigged that this meant we had the same name for both the classes and the modules. I'm quite keen to rename some of them, but I'm drawing a blank at thinking of yet another name for bits that persist stuff .oO( random name generator anyone? )

**Comment by richvdh** [comment]
wow. my earlier comment made zero sense. It's almost as if I wrote half of it, got distracted, wrote another half, got distracted again, and then just mashed the button... well done for getting the gist of what I was trying to say.

A couple of ideas about naming:
 * the whole principle of the data_store modules is that they can be connected to different databases, right? so can we call them "database"s instead?
 * or: what about calling the higher-level abstractions (of which persist_events is an example?) "persistence"s?

**Comment by erikjohnston** [comment]
Annoyingly we already have a `Database` object that is used to talk to the DB itself, though maybe that's ok? So something like: "Storage classes" -> "databases" -> "data store classes" (-> "Database")

**Comment by erikjohnston** [comment]
Ugh, the other problem with `databases` is that its going to get confused with the multiple physical databases.

**Comment by erikjohnston** [comment]
Something like "data store sets" (or "collection", or "units" or something) for layer 2? So you'd have the "main data store set" and the "state data store set"? Each of which consists of multiple classes?

**Comment by erikjohnston** [comment]
Ok, so current plan is:

- Rename `synapse.storage.data_stores` to `synapse.storage.databases`, and
- Rename `Database` to something like `DatabasePool`.

That will touch quite a bit of code/comments, so will do that as a separate PR.

**Pull request merged**